torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -287,9 +287,97 @@ void removeNopPacking(std::shared_ptr<Graph>& graph) {
     it.destroyCurrent();
+
+    if (input->outputs()[0]->uses().size() == 0) {
+      input->destroy();
+    }
+  }
+}
+
+void fixDefaultRNNState(std::shared_ptr<Graph>& graph, Node * n, int input_index) {
+  auto initial_state = n->inputs()[input_index];
+  if (initial_state->node()->kind() == onnx::Constant ||
+      (initial_state->node()->kind() == onnx::Slice &&
+       initial_state->node()->inputs()[0]->node()->kind() == onnx::Constant)) {
+    Node * shape_of_input = graph->create(onnx::Shape, 1);
+    shape_of_input->insertBefore(n);
+    shape_of_input->addInput(n->inputs()[0]);
+
+    Node * gather_indices = graph->create(onnx::Constant, 1);
+    gather_indices->insertBefore(n);
+    gather_indices->t_(attr::value, at::Scalar(1).toTensor());
+
+    Node * batch_size = graph->create(onnx::Gather, 1);
+    batch_size->insertBefore(n);
+    batch_size->addInput(shape_of_input->outputs()[0]);
+    batch_size->addInput(gather_indices->outputs()[0]);
+
+    Node * unsqueezed_batch_size = graph->create(onnx::Unsqueeze, 1);
+    unsqueezed_batch_size->insertBefore(n);
+    unsqueezed_batch_size->addInput(batch_size->outputs()[0]);
+    unsqueezed_batch_size->is_(attr::axes, {0});
+
+    Node * hidden_size = graph->create(onnx::Constant, 1);
+    hidden_size->insertBefore(n);
+    hidden_size->t_(attr::value, at::Scalar(n->i(attr::hidden_size)).toTensor());
+
+    Node * unsqueezed_hidden_size = graph->create(onnx::Unsqueeze, 1);
+    unsqueezed_hidden_size->insertBefore(n);
+    unsqueezed_hidden_size->addInput(hidden_size->outputs()[0]);
+    unsqueezed_hidden_size->is_(attr::axes, {0});
+
+    Node * num_directions = graph->create(onnx::Constant, 1);
+    num_directions->insertBefore(n);
+    num_directions->t_(attr::value, at::Scalar(n->hasAttribute(attr::direction) && n->s(attr::direction) == "bidirectional" ? 2 : 1).toTensor());
+
+    Node * unsqueezed_num_directions = graph->create(onnx::Unsqueeze, 1);
+    unsqueezed_num_directions->insertBefore(n);
+    unsqueezed_num_directions->addInput(num_directions->outputs()[0]);
+    unsqueezed_num_directions->is_(attr::axes, {0});
+
+    Node * concated_dims = graph->create(onnx::Concat, 1);
+    concated_dims->insertBefore(n);
+    concated_dims->i_(attr::axis, 0);
+    concated_dims->addInput(unsqueezed_num_directions->outputs()[0]);
+    concated_dims->addInput(unsqueezed_batch_size->outputs()[0]);
+    concated_dims->addInput(unsqueezed_hidden_size->outputs()[0]);
+
+    Node * constant_fill = graph->create(onnx::ConstantFill, 1);
+    constant_fill->insertBefore(n);
+    constant_fill->i_(attr::input_as_shape, 1);
+    constant_fill->addInput(concated_dims->outputs()[0]);
+
+    n->replaceInput(input_index, constant_fill->outputs()[0]);
+  }
+}
+
+void fixDefaultRnnHiddenState(std::shared_ptr<Graph>& graph) {
+  for (auto it = graph->nodes().begin(); it != graph->nodes().end(); ++it) {
+    auto* n = *it;
+
+    if (!isRNN(n)) {
+      continue;
+    }
+    if (n->inputs().size() < 6) {
```

Comment:
Some inline comments explaining regarding what the inputs are would be helpful (both here and below). Right now, those numbers are just some magical constants

Reply:
++ will add comment